doc: Add review document draft (#1)
dis-moi/docs: REVIEW.md

# Review process

> > This is a general process, a project specific process must be documented in its own repository.

Reviewer: ```suggestion
> This is a general process used across Dismoi by default. A project-specific process must be documented in its own repository.
```

We rely on peer review as much as possible :
> https://en.wikipedia.org/wiki/Peer_review

Try to be as benevolent as possible when reviewing your peers. 

> Someone theorise *Egoless programming* that fits our overall philosophy quite well.

Reviewer: ```suggestion
Someone theorized about *Egoless programming* which fits our overall philosophy quite well.
```

> https://en.wikipedia.org/wiki/Egoless_programming

## Branch

> When starting to work on a subject on any project, the first thing to do is creating your own branch.
> The base branch should be `master` and the new branch should reflect the subject you are working on:
> - a new feature must be prefixed by `feat/` ex: `feat/new-contribution-form`
> - a bug fix can be prefiexed by `fix/`
> Or `chore`, `docs` or any other word you see fit!

Reviewer: ```suggestion
When starting to work on a subject on any project, the first thing to do is create your own branch.
The base branch should be `main` and the new branch should reflect the subject you are working on:
- a new feature must be prefixed by `feat/` ex: `feat/new-contribution-form`
- a bug fix can be prefixed by `fix/`
Or `chore`, `docs` or any other word you see fit!
```

> > We use Conventionnal commits for commit message, and this branch naming is inspired by the keywords use there

Reviewer: ```suggestion
> We use Conventional commits for commit messages, and this branch naming is inspired by the keywords used there
```

## Commits

> For commit messages and we use Conventionnal commits convention.
> Beware, a single commit shouldn't break anything (nor the test, nor the app, nor anythign.
> We should be able to checkout any commits and run the app without any issues.

Reviewer: ```suggestion
For commit messages we use the Conventional commits convention.
Beware, a single commit shouldn't break anything (whether the test, or the app, or anything).
We should be able to check out any commits and run the app without any issues.
Also, we should try to keep commits small. Try to include only the smallest amount of change that keeps it atomic. In other words, let's create commits that only do one thing.
```

> 

Reviewer 2: Also, we try to keep commits small, and only include the smallest amount of change that keeps it atomic. In other words : let’s create commits that only do one thing.

> > When unsure just squash everyhting.

Reviewer: ```suggestion
> When unsure just squash everything.
```

## History
Keep your history clean! There can be many commits or only one, as long as the story told is consistent.
If you the history doesn't feel right, just be a negationist and use (and abuse) `git rebase -i`. 

> Just as long as you don't reright other people's history it's fine (really).

Reviewer 2: `reright` -> `overwrite`

> Just as long as you don't reright other people's history it's fine (really).

Reviewer: ```suggestion
Just as long as you don't overwrite other people's history it's fine (really).
```

> And if you feel like, feel freer to `squash` your branch history into a single commit.

Reviewer 2: Not sure I like the recommandiation, as it contradicts my recommandation suggested above …

Reviewer: Hm what's the best way to do this? Should we keep commits atomic or squash the branch history together into one big commit?

Author: Well maybe, this is just a vocabulary issue (atomic?) or maybe interpretation?

> see the discussion below regarding a "commit message guideline" :
> https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53#gistcomment-2936929

Sometimes you just a go into a commit frenzy and reach a point where you need overwrite your history. 
All those commits are related to the same feature and could be described by single simple sentence.
Feels atomic to me... or mb molecular? :D 
Well I don't really care about which is the smallest building block here. What matters is those changes could be sum up with just a few words.

One thing I'd like to avoid is having commits which don't bring any value:
```
Add `sum` function to `math.js` 
```
> even if I probably did similar things somewhere in the past

Reviewer 2: I think we agree here @lutangar 
It’s maybe just a matter of wording indeed
Let’s say that if you PR is small enough and only does one thing, you can always squash :-)

## Pull/Merge request
When ready you can open a Pull Request targeting the main branch.

> > BTW shall we rename `master` to `main`?

Reviewer: Yes!

Author: Could you create issues in the respective repositories?
I'm curious to see how you would fill in the "feature request template" for this.

OR not! You can open the PR as soon as you create your branch, 

> this can be relevant to keep other team member posted. 

Reviewer: ```suggestion
this can be relevant to keep other team members posted. 
```

> Just be sure to add the WIP keyword somewhere in the title and/or add the `WIP` label to your PR.

Reviewer: ```suggestion
Just be sure to add the WIP keyword somewhere in the title and/or add the `WIP` label to your PR (this stands for 'Work in Progress'.
```

> > Also if your PR relate to an issue in any way, don't forget to mention it (#42 for issue number 42)

Reviewer: ```suggestion
> Also if your PR relates to an issue in any way, don't forget to mention it (#42 for issue number 42)
```

> https://www.conventionalcommits.org/en/v1.0.0-beta.2/#commit-message-for-a-fix-using-an-optional-issue-number

> __Ultimate tip__ If your branch only contains one commit with a title and a body:
> ```
> doc: Add review document draft
> At last, here's the long-awaited documentation!
> ```
> The title and description of the PR will use this information.

## Reviewer - *the peer*

> Depending on the subject of you can assign non technical people as reviewers:

Reviewer: ```suggestion
Depending on the subject of the PR, you can assign non-technical people as reviewers:
```

> - member of the product to review doc, story or screeshot

Reviewer: ```suggestion
- member of the product team to review the docs, user stories or screenshots
```

> - member of the comm team to review spelling, relevance of keyword use, ...
> - member of UX UI or graphical, to review screenshot and user stories in practice, 
> > to do this the best way possible feature branches could be deployed and testable soemwhere though

Reviewer: ```suggestion
- member of the comms team to review spelling, relevance of keyword use, images and wording
- member of UX-UI or graphical team to review images, screenshots, design and user stories in practice 
> to do this in the best way possible, feature branches should be deployed and testable somewhere.
```

- At least 2 reviewer should be assigned to a PR.
- Reviewers!
  - Please be dilligent with your reviews! From our experience, PR which aren't reviewed soon tend to wait... wait.. sometime for months...even years! (I just closed a 7 years old PR recently)
  - Tip: If you don't understand anything you're reading, you're might not be the right person for the job! And that's OK! 
  Just let us know and we'll help you understand or choose another person fitting the task (at least try to look for typo:))

>   - We love our work to be read by others (who said egoless?!) so please try to show you care and leave some comments in the diff!

Reviewer 2: ♥

> - At least 2 reviewer should be assigned to a PR.
> - Reviewers!
>   - Please be dilligent with your reviews! From our experience, PR which aren't reviewed soon tend to wait... wait.. sometime for months...even years! (I just closed a 7 years old PR recently)
>   - Tip: If you don't understand anything you're reading, you're might not be the right person for the job! And that's OK! 
>   Just let us know and we'll help you understand or choose another person fitting the task (at least try to look for typo:))
>   - We love our work to be read by others (who said egoless?!) so please try to show you care and leave some comments in the diff!

Reviewer: ```suggestion
- At least 2 reviewers should be assigned to a PR.
- Reviewers!
  - Please be diligent with your reviews! From our experience, PR which aren't reviewed quickly tend to wait... wait.. sometimes for months...even years! (I just closed a 7 year old PR recently)
  - Tip: If you don't understand anything you're reading, you might not be the right person for the job! And that's OK! 
  Just let us know and we'll help you understand or choose another person suited to the task (at least try to look for typos :) )
  - We love our work to be read by others (who said egoless?!) so please try to show you care and leave some comments in the diff!
```

- A PR should be merged in the next 72h? at most

> - You must have at least one approval to merge a PR

Reviewer 2: Paragraph proposal : 

## Assignment / Merge

We consider that the PR author is responsible for it during its lifecycle. It means that the author :
 - decides who to require a review from (but everyone can review/comment)
 - decides when he got sufficient feedbacks
 - decides whether to handle or not non-blocking comments
 - decides when to merge, and triggers it
 
 But then again, even if he has the final saying, we usually try to work a team, discuss things and find consensus.

Reviewer: I like this, it's really helpful for the PR author to know that they are the one responsible for their PR.

Just so I understand, the PR author does *not* have the final say on merging something to main, correct? That responsibility should be with one of the reviewers right? I'm just thinking of a potential case where a PR author wants to merge but there are clear issues with the PR. In that case, it's best if someone with knowledge of the system makes the final call. What do you think?

Author: Discussed with @felix-lambert last week, and what he suggested kind of make sense:
- if the reviewer approved the PR, he should be the one to merge it

> Note: it avoids one extra cycle whereas the PR is approved and can be merged

> Note 2: Actually we kind of did that already but not always, I'm not really sure of the unwritten rule behind this.

But what about when there are 2 reviewers or more then? Is one approval still sufficient?
Shall it be a rule? An exception?

Reviewer 2: > But what about when there are 2 reviewers or more then

- That’s why I suggested that the PR author is the one in position to know if he got enough feedback from review(s) or not, and decide to merge. Even if I get how it adds a «cycle» …

- It also solves the problem of considering if the remaining discussions are blocking or not

Of course, this only works assuming that PR author is experienced enough to be able make this decision.

In fact, we could adopt my proposal if the PR author is the maintainer, 
But if the author is not a maintainer, different rule : Any maintainer can merge.

### Unwritten rules (so far)
Mind the state of the review :
- approve
- comment
- request changes

> If the PR is approved, it's approved period, even if there are comments left, its just mean it's not these notices aren't that important after all.

Reviewer 2: typo+reword =>  it just means these comments are not blocking, and may be taken care of later.

> If the PR is approved, it's approved period, even if there are comments left, its just mean it's not these notices aren't that important after all.

Reviewer: ```suggestion
If the PR is approved, it's approved, period. Even if there are comments left, it just means these comments are not blocking, and may be taken care of later.
```

> If one of the reviewer requested changes, changes must be applied or he must convinced otherwise until changes are approved. 
> And then request a re-review (there is a button for that)
> 
> ## Resources
> There are plenty of excellent resource out there feel freee to add anything relevant here
> 
> ## Tools
> Any tools that could help to be a git, gihub power user?

Reviewer: ```suggestion
If one of the reviewers requests changes, the changes must be applied or the reviewer must be convinced otherwise until the PR is approved.
Once changes are applied, the PR author should then request a re-review (there is a button for that).

## Resources
There are plenty of excellent resources out there, so feel free to add anything relevant here.

## Tools
Any tools that could help folks to become a Git, Github power user?

- [Managing your work on GitHub](https://docs.github.com/en/github/managing-your-work-on-github)
- [Using Git](https://docs.github.com/en/github/using-git)
- [A Step-by-Step Guide to Git](https://opensource.com/article/18/1/step-step-guide-git)
- [How to become a Git expert](https://www.freecodecamp.org/news/how-to-become-a-git-expert-e7c38bf54826/)
```
